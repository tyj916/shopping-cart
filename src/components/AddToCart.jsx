import { useState } from "react";

function AddToCart() {
  const [count, setCount] = useState(1);

  function handleCountChange(e) {
    setCount(+e.target.value);
  }

  function addCount() {
    setCount(count + 1);
  }

  function minusCount() {
    if (count > 1) {
      setCount(count - 1);
    }
  }

  return (
    <div className='add-to-cart-controller'>
      <span className='quantity-controller'>
        <button type='button' onClick={minusCount}>-</button>
        <input type="tel" aria-label='count' value={count} onChange={handleCountChange} />
        <button type='button' onClick={addCount}>+</button>
      </span>
      <button type='button'>Add to cart</button>
    </div>
  )
}

export default AddToCart;