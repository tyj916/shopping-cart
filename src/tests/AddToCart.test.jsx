import { render, screen } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import AddToCart from "../components/AddToCart";
import userEvent from "@testing-library/user-event";

describe('Add to cart component', () => {
  it('renders add to cart button with number control', () => {
    const { container } = render(<AddToCart />);
    expect(container).toMatchSnapshot();
  });

  it('renders value 2 after add button is clicked', async () => {
    const user = userEvent.setup();

    render(<AddToCart />);
    const button = screen.getByRole('button', { name: '+' });

    await user.click(button);

    expect(+screen.getByLabelText('quantity').value).toBe(2);
  });

  it('should minus 1 quantity after minus button is clicked', async () => {
    const user = userEvent.setup();

    render(<AddToCart />);
    const addButton = screen.getByRole('button', { name: '+' });
    const minusButton = screen.getByRole('button', { name: '-' });
    const input = screen.getByLabelText('quantity');

    await user.click(addButton);
    await user.click(minusButton);

    expect(+input.value).toBe(1);
  });

  it('should not lower than 1', async () => {
    const user = userEvent.setup();

    render(<AddToCart />);
    const minusButton = screen.getByRole('button', { name: '-' });
    const input = screen.getByLabelText('quantity');

    await user.click(minusButton);

    expect(+input.value).toBe(1);
  });
});