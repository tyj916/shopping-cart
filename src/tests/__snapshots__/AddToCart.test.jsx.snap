// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Add to cart component > renders add to cart button with number control 1`] = `
<div>
  <div
    class="add-to-cart-controller"
  >
    <span
      class="quantity-controller"
    >
      <button
        type="button"
      >
        -
      </button>
      <input
        aria-label="count"
        type="tel"
        value="1"
      />
      <button
        type="button"
      >
        +
      </button>
    </span>
    <button
      type="button"
    >
      Add to cart
    </button>
  </div>
</div>
`;
